File: src/main/scala/com/hoi4utils/ui/CustomTooltipController.scala
package com.hoi4utils.ui

import com.hoi4utils.hoi4.tooltip.CustomTooltip
import com.hoi4utils.ui.javafx_ui.table.TableViewWindow
import com.hoi4utils.{HOIIVFiles, HOIIVUtils}
import javafx.fxml.{FXML, Initializable}
import javafx.scene.control.*
import scalafx.collections.ObservableBuffer

import java.io.File
import java.net.URL
import java.util.ResourceBundle
import scala.compiletime.uninitialized

/// // * todo: have to redo some functionality to work with new localization system
class CustomTooltipController extends HOIIVUtilsAbstractController with TableViewWindow with Initializable {
  
  @FXML var idVersion: Label = uninitialized
  @FXML var tooltipIdTableColumn: TableColumn[CustomTooltip,String] = uninitialized
  @FXML var tooltipTextTableColumn: TableColumn[CustomTooltip,String] = uninitialized
  @FXML var tooltipFileComboBox: ComboBox[File]          = uninitialized
  @FXML var tooltipLocalizationFileComboBox: ComboBox[File] = uninitialized
  @FXML var tooltipFileBrowseButton: Button              = uninitialized
  @FXML var tooltipLocalizationFileBrowseButton: Button  = uninitialized
  @FXML var customTooltipTableView: TableView[CustomTooltip] = uninitialized

  private var tooltipFile: Option[File] = None 
  
  /* default */
  setFxmlResource("CustomTooltip.fxml")
  setTitle("Custom Tooltips")

  // ScalaFX‐friendly backing list
  private val customTooltipBuf: ObservableBuffer[CustomTooltip] = ObservableBuffer.empty
  
  override def initialize(location: URL, resources: ResourceBundle): Unit = {
    idVersion.setText(HOIIVUtils.get("version"))
    // wire up the JavaFX TableView using your existing helper:
    loadTableView(this, customTooltipTableView, customTooltipBuf, CustomTooltip.dataFunctions())
  }

  // --- action handlers ---
  @FXML def handleTooltipFileBrowseAction(): Unit = {
    val initial = HOIIVFiles.Mod.common_folder
    val selectedFile = JavaFXUIManager.openChooser(tooltipFileBrowseButton, initial, false)
    if (selectedFile != null) {
      tooltipFileComboBox.setValue(selectedFile)
      tooltipFile = Some(selectedFile)
      CustomTooltip.loadTooltips(selectedFile)
      CustomTooltip.getTooltips.foreach { tooltip =>
        customTooltipBuf += tooltip
      }
    }
    
  }

  @FXML def handleTooltipLocalizationFileBrowseAction(): Unit = {
    // todo 
  }

  // TableViewWindow stub
  override def setDataTableCellFactories(): Unit = {
  }
}

File: src/main/scala/com/hoi4utils/ui/hoi4localization/CustomTooltipController.java
//package com.hoi4utils.ui.hoi4localization;
//
//import com.hoi4utils.HOIIVUtils;
//import com.hoi4utils.HOIIVFiles;
//import com.hoi4utils.hoi4.tooltip.CustomTooltip;
//import com.hoi4utils.hoi4.tooltip.CustomTooltip$;
//import com.hoi4utils.ui.JavaFXUIManager;
//import com.hoi4utils.ui.HOIIVUtilsAbstractController;
//import com.hoi4utils.ui.javafx.table.TableViewWindow;
//import javafx.collections.FXCollections;
//import javafx.collections.ObservableList;
//import javafx.fxml.FXML;
//import javafx.scene.control.*;
//import scala.Function1;
//import scala.collection.immutable.List;
//import scala.jdk.javaapi.CollectionConverters;
//
//import java.io.File;
//import java.util.ArrayList;
//import java.util.Collection;
//
///**
// * todo: have to redo some functionality to work with new localization system
// */
//public class CustomTooltipController extends HOIIVUtilsAbstractController implements TableViewWindow {
//
//	@FXML
//	public Label idVersion;
//	@FXML
//	TableColumn<CustomTooltip, String> tooltipIdTableColumn;
//	@FXML
//	TableColumn<CustomTooltip, String> tooltipTextTableColumn;
//	@FXML
//	ComboBox<File> tooltipFileComboBox;
//	@FXML
//	ComboBox<File> tooltipLocalizationFileComboBox;
//	@FXML
//	Button tooltipFileBrowseButton;
//	@FXML
//	Button tooltipLocalizationFileBrowseButton;
//	@FXML
//	TableView<CustomTooltip> customTooltipTableView;
//	private File tooltipFile;
//
//	private ObservableList<CustomTooltip> customTooltipList;
//
//	public CustomTooltipController() {
//		/* window */
//		setFxmlResource("CustomTooltip.fxml");
//		setTitle("HOIIVUtils Custom ToolTip Localization Window");
//
//		customTooltipList = FXCollections.observableArrayList();
//	}
//
//	/**
//	 * {@inheritDoc}
//	 *
//	 */
//	@FXML
//	void initialize() {
//		includeVersion();
//
//		/* table */
//		loadTableView(this, customTooltipTableView, customTooltipList, CustomTooltip.dataFunctions());
//	}
//
//	private void includeVersion() {
//		idVersion.setText(HOIIVUtils.get("version").toString());
//	}
//
//	@Override
//	public void setDataTableCellFactories() {
//	}
//
//	/**
//	 * @param customTooltipList
//	 */
//	@Deprecated
//	public void setCustomTooltipList(ObservableList<CustomTooltip> customTooltipList) {
//		this.customTooltipList = customTooltipList;
//	}
//
//	/**
//	 * @param customTooltips
//	 */
//	@Deprecated
//	public void setCustomTooltipList(Collection<CustomTooltip> customTooltips) {
//		this.customTooltipList = FXCollections.observableArrayList();
//		this.customTooltipList.addAll(customTooltips);
//	}
//
//	public void updateCustomTooltipList(Collection<CustomTooltip> customTooltips) {
//		this.customTooltipList.clear();
//		this.customTooltipList.addAll(customTooltips);
//	}
//
//	public void addCustomTooltips(Collection<CustomTooltip> customTooltips) {
//		this.customTooltipList.addAll(customTooltips);
//	}
//
//	/* action handlers */
//	public void handleTooltipFileBrowseAction() {
//		File initialFocusDirectory = HOIIVFiles.Mod.common_folder;
//		File selectedFile = JavaFXUIManager.openChooser(tooltipFileBrowseButton, initialFocusDirectory, false);
//		System.out.println(selectedFile);
//		if (selectedFile != null) {
//			tooltipFileComboBox.setValue(selectedFile); // selectedFile.getAbsolutePath()
//			tooltipFile = selectedFile;
//		}
//
//		/* load custom tooltips from selected file */
//		CustomTooltip.loadTooltips(tooltipFile);
//		ArrayList<CustomTooltip> tooltips = CustomTooltip.getTooltips();
//		if (tooltips == null) {
//			System.out.println("No custom tooltips found");
//			return;
//		}
//		updateCustomTooltipList(tooltips);
//	}
//
//	public void handleTooltipLocalizationFileBrowseAction() {
////		File initialFocusLocDirectory = HOIIVUtilsFiles.mod_localization_folder;
////		File selectedFile = openChooser(tooltipLocalizationFileBrowseButton, initialFocusLocDirectory, false);
////		
////		System.out.println(selectedFile);
////		
////		if (selectedFile != null) {
////			tooltipLocalizationFileComboBox.setValue(selectedFile); // selectedFile.getAbsolutePath()
////			try {
////				localizationFile = new LocalizationFile(selectedFile);
////			} catch (IllegalLocalizationFileTypeException e) {
////				e.printStackTrace();
////				return;
////			}
////		}
////
////		/* load custom tooltip associated localization from selected file */
////		if (tooltipFile == null) {
////			return;
////		}
////		localizationFile.read();
////
////		for (CustomTooltip tooltip : customTooltipList) {
////			String tooltipID = tooltip.id();
////			Localization tooltipLocalization = localizationFile.getLocalization(tooltipID);
////			if (tooltipLocalization == null) {
////				return;
////			}
////
////			tooltip.setLocalization(tooltipLocalization);
////		}
////
////		customTooltipTableView.refresh(); // this is important to ensure the table is refreshed when the list data
////											// is modified, otherwise the localization won't appear in the table.
//	}
//}
